src/dataflow/src/render/mod.rs

```diff
@@ -674,26 +674,67 @@ pub(crate) fn build_dataflow<A: Allocate>(
                     ))
                     .expect("Sink source collection not loaded");
 
-                // TODO(frank): consolidation is only required for a collection,
-                // not for arrangements. We can perform a more complicated match
-                // here to determine which case we are in to avoid this call.
-                use differential_dataflow::operators::consolidate::Consolidate;
-                let collection = collection.consolidate();
+                // TODO(frank): Consolidation can be avoided for arranged inputs.
+
+                // These steps perform consolidation, but do not devolve the stream
+                // from batches to individual updates. That can happen for sinks that
+                // require e.g. updates shuffled, but for sinks that do not require
+                // such it is much more efficient to work off of shared batches.
+                use differential_dataflow::trace::implementations::ord::{
+                    OrdKeyBatch, OrdKeySpine,
+                };
+                let batches: timely::dataflow::Stream<
+                    _,
+                    std::rc::Rc<OrdKeyBatch<Row, Timestamp, isize>>,
+                > = collection
+                    .map(|k| (k, ()))
+                    .arrange_named::<OrdKeySpine<_, _, _>>("SinkConsolidation")
+                    .stream;
```

Comment:
this might be a dumb question but in a case like this it seems like we don't want the trace at all so is there any way to avoid maintaining it? or do we somehow already do that?

Reply:
Differential dataflow is one step ahead of you: by retaining only the stream and dropping the trace, it will not be maintained.

Reply:
nice :D 

how does that happen btw -- like i feel like there is a magical interaction between the rust type system / the rust move and borrow checker and differential thats going on here but I can't explain it

Reply:
The trace is reference counted, and the `arrange` operator only holds a weak reference to it. If all strong references are dropped, the trace vanishes. The operator needs to explicitly upgrade its reference each time it runs, and understands that if it cannot, it does not need to fold any new batches in to a trace.

Reply:
got it that makes a ton of sense